src/io/flutter/editor/InlinePreviewViewController.java

```diff
@@ -0,0 +1,398 @@
+          lastScreenshotBoundsWindow.translate(-visibleRect.x, -visibleRect.y);
+          lockUpdate = true;
+        }
+      }
+
+      if (!lockUpdate){
+        lastLockedRectangle = null;
+        if (start.y <= visibleEnd && endY >= visibleStart) {
+          if (visibleStart > previewStartY) {
+            previewStartY = Math.max(previewStartY, visibleStart);
+            previewStartY = Math.min(previewStartY, Math.min(endY - previewHeight, visibleEnd - previewHeight));
+          }
+          screenshotBounds = new Rectangle(previewStartX, previewStartY, previewWidth, previewHeight);
+          lastScreenshotBoundsWindow = new Rectangle(screenshotBounds);
+          lastScreenshotBoundsWindow.translate(-visibleRect.x, -visibleRect.y);
+        }
+      }
+    }
+
+    if (popupActive()) {
+      lastLockedRectangle = new Rectangle(visibleRect);
+    }
+  }
+
+  @Override
+  protected Dimension getPreviewSize() {
+    int previewWidth;
+    int previewHeight;
+    previewWidth = PREVIEW_MAX_WIDTH;
+    previewHeight = PREVIEW_MAX_HEIGHT;
+
+    if (getDescriptor() == null) {
+      previewWidth = Math.round(previewWidth * previewWidthScale);
+      previewHeight = Math.round(previewHeight * previewWidthScale);
+    }
+    return new Dimension(previewWidth, previewHeight);
+  }
+
+  private void updateVisibleArea(Rectangle newRectangle) {
+    visibleRect = newRectangle;
+    if (getDescriptor() == null || getData().getMarker() == null) {
+      if (!visible) {
+        visible = true;
+        onVisibleChanged();
+      }
+      return;
+    }
+    final TextRange marker = getData().getMarker();
+    if (marker == null) return;
+
+    final Point start = offsetToPoint(marker.getStartOffset());
+    final Point end = offsetToPoint(marker.getEndOffset());
+    final boolean nowVisible = newRectangle == null || newRectangle.y <= end.y && newRectangle.y + newRectangle.height >= start.y ||
+                               updateVisiblityLocked(newRectangle);
+    if (visible != nowVisible) {
+      visible = nowVisible;
+      onVisibleChanged();
+    }
+  }
+
+  @Override
+  public void dispose() {
+  }
+
+  @Override
+  protected Component getComponent() {
+    return getEditor().getComponent();
+  }
+
+  @Override
+  protected VirtualFile getVirtualFile() {
+    return getEditor().getVirtualFile();
+  }
+
+  @Override
+  protected Document getDocument() {
+    return getEditor().getDocument();
+  }
+
+  @Override
+  protected void showPopup(Point location, DiagnosticsNode node) {
+    location =
+        SwingUtilities.convertPoint(
+          getEditor().getContentComponent(),
+          location,
+          getEditor().getComponent()
+        );
+    popup = PropertyEditorPanel
+      .showPopup(data.context.inspectorGroupManagerService, getEditor(), node, node.getCreationLocation().getLocation(), data.context.flutterDartAnalysisService, location);
+  }
+
+  @Override
+  TextRange getActiveRange() {
```

Comment:
I assume null return is okay?

Reply:
it is. annotated the base class method as @Nullable to get warnings.